src/kvp/mod.rs

```diff
@@ -0,0 +1,546 @@
+        let value = V::from_str(value.1.as_ref()).context(InvalidValueSnafu)?;
+
+        Ok(Self { key, value })
+    }
+}
+
+impl<V> Display for KeyValuePair<V>
+where
+    V: ValueExt,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}={}", self.key, self.value)
+    }
+}
+
+impl<V> Serialize for KeyValuePair<V>
+where
+    V: ValueExt,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(1))?;
+        map.serialize_entry(&self.key, &self.value)?;
+        map.end()
+    }
+}
+
+struct KeyValuePairVisitor<V> {
+    marker: PhantomData<V>,
+}
+
+impl<'de, V> Visitor<'de> for KeyValuePairVisitor<V>
+where
+    V: Deserialize<'de> + ValueExt + Default,
+{
+    type Value = KeyValuePair<V>;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a valid key/value pair (label or annotation)")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::MapAccess<'de>,
+    {
+        if let Some((key, value)) = map.next_entry()? {
+            return Ok(KeyValuePair::new(key, value));
+        }
+
+        Err(serde::de::Error::custom("expected at least one map entry"))
+    }
+}
+
+impl<'de, V> Deserialize<'de> for KeyValuePair<V>
+where
+    V: Deserialize<'de> + ValueExt + Default,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(KeyValuePairVisitor {
+            marker: PhantomData,
+        })
+    }
+}
+
+impl<V> KeyValuePair<V>
+where
+    V: ValueExt,
+{
+    /// Creates a new [`KeyValuePair`] from a validated [`Key`] and value.
+    pub fn new(key: Key, value: V) -> Self {
+        Self { key, value }
+    }
+
+    /// Returns an immutable reference to the pair's [`Key`].
+    pub fn key(&self) -> &Key {
+        &self.key
+    }
+
+    /// Returns an immutable reference to the pair's value.
+    pub fn value(&self) -> &V {
+        &self.value
+    }
+}
+
+#[derive(Debug, Snafu)]
+pub enum KeyValuePairsError<E>
+where
+    E: std::error::Error + 'static,
+{
+    #[snafu(display("key/value pair already present"))]
+    AlreadyPresent,
+
+    #[snafu(display("failed to parse key/value pair"))]
+    KeyValuePairParse { source: KeyValuePairError<E> },
+}
+
+/// [`KeyValuePairs`] is a list of [`KeyValuePair`]. This collection **doesn't**
+/// provide any de-duplication mechanism, meaning multiple [`KeyValuePair`]s
+/// with the same content can be present at the same time. However, converting
+/// to a [`BTreeMap<String, String>`] removes any duplicate data. Order matters
+/// in this case: later labels overwrite previous onces.
+#[derive(Clone, Debug, Default)]
+pub struct KeyValuePairs<V: ValueExt>(BTreeSet<KeyValuePair<V>>);
```

Comment:
Couldn't this be a `BTreeMap<K, V>`? Why does this need a dedicated type anyway?

Reply:
I mainly wanted to have a controlled way of creating a list of `KeyValuePair`. One example is `Labels::recommended()`.

Reply:
Couldn't that be `kvp::label::recommended()`?

Reply:
Which would return a `BTreeMap<Key, LabelValue>` or a `BTreeSet<KeyValuePair<LabelValue>>`?

Reply:
`BTreeMap`. `BTreeSet` doesn't have the semantics you want. For another example, saying `a=b` always takes precedence over `a=a` is just bizarre.

Reply:
Or... we could just use the collection directly, and not reinvent the wheel.

Reply:
@NickLarsenNZ and I decided to move forward as is. We can refine the feature in later PRs. I already have some stuff in mind :)